### PR TITLE
Run plugin hooks in order determined by dependencies

### PR DESCRIPTION
**Pull request by kmatuszkiewicz**
### Description
DobbyRdkPluginManager now processes RDK plugins for a given hook according to the order of plugin dependency. 
Before this change, RDK plugins were processed according to the order in which they are declared in the "RdkPlugins" section of the "defs-plugins" schema file. This can result in hook execution failures when plugin A which depends on plugin B is processed before plugin B. 

After the change, if plugin A depends on plugin B, hooks for plugin B will be run before hooks for plugin A.
This is inverted (A before B) for postHalt and postStop so that the plugins on which other plugins depend are shut down later.

The dependencies can be defined in the container spec. Each plugin spec now contains a "dependsOn" array.
If, for example, plugin A depends on plugins B and C, then plugin A's spec can look like this:
"a": {
    "required": false,
    "dependsOn": [ "b", "c" ],
    "data": {}
}

Dependency cycles are detected and make DobbyRdkPluginManager abort before it starts running the plugin hooks.

### Test Procedure
Create a container spec with "dependsOn" clauses defined for some plugins, use DobbyPluginLauncher to check if hooks are executed according to the order of dependency.

### Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other (doesn't fit into the above categories - e.g. documentation updates)

### Requires Bitbake Recipe changes?
- [ ] The base Bitbake recipe (`meta-rdk-ext/recipes-containers/dobby/dobby.bb`) must be modified to support the changes in this PR (beyond updating `SRC_REV`)

### CODE REVIEW

File: pluginLauncher/lib/source/DobbyRdkPluginManager.cpp [hunk 1]
```diff
@@ -578,17 +666,23 @@ bool DobbyRdkPluginManager::runPlugins(const IDobbyRdkPlugin::HintFlags &hookPoi
...
 
-    const auto pluginsInConfig = mContainerConfig->rdk_plugins->names_of_plugins;
-    const int rdkPluginCount = mContainerConfig->rdk_plugins->plugins_count;
-
-    std::string pluginName;
+    // Determine the order of launching based on the dependencies.
+    std::vector<std::string> launchOrder;
+    if (hookPoint < IDobbyRdkPlugin::HintFlags::PostHaltFlag)
+    {
+        launchOrder = mDependencySolver.getOrderOfDependency();
```
Comment (TeknoVenus) [inline pluginLauncher/lib/source/DobbyRdkPluginManager.cpp:673]: `getOrderOfDependency()` can return an empty vector if there's a circular dependency or other error. There's no check for an empty vector here so `runPlugins()` will still return true in that scenario. `runPlugins()` should probably return false if a cycle is detected

File: pluginLauncher/lib/source/DobbyRdkPluginManager.cpp [hunk 2]
```diff
@@ -296,6 +297,74 @@ void DobbyRdkPluginManager::loadPlugins()
...
+    // Check if required plugins are loaded, store plugin dependencies.
+    for (size_t i = 0; i < rdkPluginCount; i++)
+    {
+        const std::string pluginName = pluginsInConfig[i];
+        if (!isLoaded(pluginName))
+        {
+            if (isRequired(pluginName))
+            {
+                AI_LOG_ERROR_EXIT("Required plugin %s isn't loaded, but present in the container config - aborting", pluginName.c_str());
+                return false;
```
Comment (TeknoVenus) [inline pluginLauncher/lib/source/DobbyRdkPluginManager.cpp:341]: This method returns false but the return value is never used/checked as this is called from the constructor.

Potential idea - add an `isValid` flag that is set to `false` in these scenarios (and possibly during the `loadPlugins()` method as well, then other methods can check this flag to check if plugin loading & pre-processing was successful and fail fast if not.

File: pluginLauncher/lib/source/DobbyRdkPluginManager.cpp [hunk 3]
```diff
@@ -578,17 +666,23 @@ bool DobbyRdkPluginManager::runPlugins(const IDobbyRdkPlugin::HintFlags &hookPoi
...
     // Run all the plugins
-    for (size_t i = 0; i < rdkPluginCount; i++)
+    for (const std::string &pluginName : launchOrder)
     {
-        pluginName = pluginsInConfig[i];
-
-        const bool required = mContainerConfig->rdk_plugins->required_plugins[i];
+        const bool required = isRequired(pluginName);
 
         if (required && !isLoaded(pluginName))
```
Comment (TeknoVenus) [inline pluginLauncher/lib/source/DobbyRdkPluginManager.cpp:687]: Do we still need all these checks here? Some of this is duplicated in the `preProcess` section now.

**Review by TeknoVenus — CHANGES_REQUESTED** [review]
If the OCI bundle contains an `rdkPlugins` section where a plugin does not contain a `dependsOn` array, DobbyDaemon will segfault.
```
0000006293.318538 NFO: < M:DobbyRdkPluginManager.cpp F:loadPlugins L:288 > Loaded plugin 'logging' from '/usr/lib/plugins/dobby/libLoggingPlugin.so.1'
==19809== Invalid read of size 8
==19809==    at 0xC11A212: LoggingPlugin::getDependencies[abi:cxx11]() const (LoggingPlugin.cpp:108)
==19809==    by 0x641BCB: DobbyRdkPluginManager::preprocessPlugins() (DobbyRdkPluginManager.cpp:351)
==19809==    by 0x64071A: DobbyRdkPluginManager::DobbyRdkPluginManager(std::shared_ptr<rt_dobby_schema>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::shared_ptr<DobbyRdkPluginUtils> const&) (DobbyRdkPluginManager.cpp:61)
==19809==    by 0x543C62: void __gnu_cxx::new_allocator<DobbyRdkPluginManager>::construct<DobbyRdkPluginManager, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&>(DobbyRdkPluginManager*, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&) (in /usr/sbin/DobbyDaemon)
==19809==    by 0x542635: void std::allocator_traits<std::allocator<DobbyRdkPluginManager> >::construct<DobbyRdkPluginManager, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&>(std::allocator<DobbyRdkPluginManager>&, DobbyRdkPluginManager*, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&) (alloc_traits.h:475)
==19809==    by 0x540CF2: std::_Sp_counted_ptr_inplace<DobbyRdkPluginManager, std::allocator<DobbyRdkPluginManager>, (__gnu_cxx::_Lock_policy)2>::_Sp_counted_ptr_inplace<std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&>(std::allocator<DobbyRdkPluginManager>, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&) (shared_ptr_base.h:520)
==19809==    by 0x53E8CB: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<DobbyRdkPluginManager, std::allocator<DobbyRdkPluginManager>, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&>(std::_Sp_make_shared_tag, DobbyRdkPluginManager*, std::allocator<DobbyRdkPluginManager> const&, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&) (shared_ptr_base.h:615)
==19809==    by 0x53CBDE: std::__shared_ptr<DobbyRdkPluginManager, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<std::allocator<DobbyRdkPluginManager>, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&>(std::_Sp_make_shared_tag, std::allocator<DobbyRdkPluginManager> const&, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&) (shared_ptr_base.h:1100)
==19809==    by 0x53AE75: std::shared_ptr<DobbyRdkPluginManager>::shared_ptr<std::allocator<DobbyRdkPluginManager>, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&>(std::_Sp_make_shared_tag, std::allocator<DobbyRdkPluginManager> const&, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&) (shared_ptr.h:319)
==19809==    by 0x5392FD: std::shared_ptr<DobbyRdkPluginManager> std::allocate_shared<DobbyRdkPluginManager, std::allocator<DobbyRdkPluginManager>, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&>(std::allocator<DobbyRdkPluginManager> const&, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&) (shared_ptr.h:620)
==19809==    by 0x5374CB: std::shared_ptr<DobbyRdkPluginManager> std::make_shared<DobbyRdkPluginManager, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&>(std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&) (shared_ptr.h:636)
==19809==    by 0x530201: DobbyManager::startContainerFromBundle(ContainerId const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::list<int, std::allocator<int> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) (DobbyManager.cpp:818)
```
To reproduce - use DobbyBundleGenerator to convert `sleepy.json` spec to OCI bundle:
```
DobbyBundleGenerator -i ./srcDobby/tests/dobby_specs/sleepy.json -o ./sleepybundle
```
and attempt to run - DobbyDaemon will crash

File: pluginLauncher/lib/source/DobbyRdkPluginDependencySolver.cpp [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+ *
+ * @return True if the dependency has been added successfully,
+ * false otherwise (one of the plugins had not been added to the solver).
+ *
+ */
+bool DobbyRdkPluginDependencySolver::addDependency(const std::string &pluginName, const std::string &dependencyName)
+{
+    AI_LOG_FN_ENTRY();
+
+    if (mDescriptorMap.count(pluginName) == 0)
```
Comment (TeknoVenus) [inline pluginLauncher/lib/source/DobbyRdkPluginDependencySolver.cpp:80]: These checks are case sensitive - plugin names should be case insensitive. Convert `pluginName` to lowercase before checking

**Comment by kmatuszkiewicz** [comment]
> If the OCI bundle contains an `rdkPlugins` section where a plugin does not contain a `dependsOn` array, DobbyDaemon will segfault.
> 
> ```
> 0000006293.318538 NFO: < M:DobbyRdkPluginManager.cpp F:loadPlugins L:288 > Loaded plugin 'logging' from '/usr/lib/plugins/dobby/libLoggingPlugin.so.1'
> ==19809== Invalid read of size 8
> ==19809==    at 0xC11A212: LoggingPlugin::getDependencies[abi:cxx11]() const (LoggingPlugin.cpp:108)
> ==19809==    by 0x641BCB: DobbyRdkPluginManager::preprocessPlugins() (DobbyRdkPluginManager.cpp:351)
> ==19809==    by 0x64071A: DobbyRdkPluginManager::DobbyRdkPluginManager(std::shared_ptr<rt_dobby_schema>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::shared_ptr<DobbyRdkPluginUtils> const&) (DobbyRdkPluginManager.cpp:61)
> ==19809==    by 0x543C62: void __gnu_cxx::new_allocator<DobbyRdkPluginManager>::construct<DobbyRdkPluginManager, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&>(DobbyRdkPluginManager*, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&) (in /usr/sbin/DobbyDaemon)
> ==19809==    by 0x542635: void std::allocator_traits<std::allocator<DobbyRdkPluginManager> >::construct<DobbyRdkPluginManager, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&>(std::allocator<DobbyRdkPluginManager>&, DobbyRdkPluginManager*, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&) (alloc_traits.h:475)
> ==19809==    by 0x540CF2: std::_Sp_counted_ptr_inplace<DobbyRdkPluginManager, std::allocator<DobbyRdkPluginManager>, (__gnu_cxx::_Lock_policy)2>::_Sp_counted_ptr_inplace<std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&>(std::allocator<DobbyRdkPluginManager>, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&) (shared_ptr_base.h:520)
> ==19809==    by 0x53E8CB: std::__shared_count<(__gnu_cxx::_Lock_policy)2>::__shared_count<DobbyRdkPluginManager, std::allocator<DobbyRdkPluginManager>, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&>(std::_Sp_make_shared_tag, DobbyRdkPluginManager*, std::allocator<DobbyRdkPluginManager> const&, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&) (shared_ptr_base.h:615)
> ==19809==    by 0x53CBDE: std::__shared_ptr<DobbyRdkPluginManager, (__gnu_cxx::_Lock_policy)2>::__shared_ptr<std::allocator<DobbyRdkPluginManager>, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&>(std::_Sp_make_shared_tag, std::allocator<DobbyRdkPluginManager> const&, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&) (shared_ptr_base.h:1100)
> ==19809==    by 0x53AE75: std::shared_ptr<DobbyRdkPluginManager>::shared_ptr<std::allocator<DobbyRdkPluginManager>, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&>(std::_Sp_make_shared_tag, std::allocator<DobbyRdkPluginManager> const&, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&) (shared_ptr.h:319)
> ==19809==    by 0x5392FD: std::shared_ptr<DobbyRdkPluginManager> std::allocate_shared<DobbyRdkPluginManager, std::allocator<DobbyRdkPluginManager>, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&>(std::allocator<DobbyRdkPluginManager> const&, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&) (shared_ptr.h:620)
> ==19809==    by 0x5374CB: std::shared_ptr<DobbyRdkPluginManager> std::make_shared<DobbyRdkPluginManager, std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&>(std::shared_ptr<rt_dobby_schema>&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, char const (&) [23], std::shared_ptr<DobbyRdkPluginUtils>&) (shared_ptr.h:636)
> ==19809==    by 0x530201: DobbyManager::startContainerFromBundle(ContainerId const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::list<int, std::allocator<int> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&) (DobbyManager.cpp:818)
> ```
> 
> To reproduce - use DobbyBundleGenerator to convert `sleepy.json` spec to OCI bundle:
> 
> ```
> DobbyBundleGenerator -i ./srcDobby/tests/dobby_specs/sleepy.json -o ./sleepybundle
> ```
> 
> and attempt to run - DobbyDaemon will crash

There was no segfault. False positive due to faulty test environment.

File: pluginLauncher/lib/source/DobbyRdkPluginManager.cpp (hunk 1 above)
Comment (kmatuszkiewicz) [reply pluginLauncher/lib/source/DobbyRdkPluginManager.cpp:678]: Agreed, done.

File: pluginLauncher/lib/source/DobbyRdkPluginManager.cpp (hunk 2 above)
Comment (kmatuszkiewicz) [reply pluginLauncher/lib/source/DobbyRdkPluginManager.cpp:349]: Good idea, I've added the mValid flag.

File: pluginLauncher/lib/source/DobbyRdkPluginManager.cpp (hunk 3 above)
Comment (kmatuszkiewicz) [reply pluginLauncher/lib/source/DobbyRdkPluginManager.cpp]: Agreed, removed.

File: pluginLauncher/lib/source/DobbyRdkPluginDependencySolver.cpp (hunk 4 above)
Comment (kmatuszkiewicz) [reply pluginLauncher/lib/source/DobbyRdkPluginDependencySolver.cpp]: Done.

**Review by TeknoVenus — APPROVED** [review]
LGTM!

**Pull request merged**